models/balancer/arbitrum/balancer_v2_arbitrum_bpt_prices.sql

```diff
@@ -1,66 +1,160 @@
 {{
     config(
         schema = 'balancer_v2_arbitrum',
-        alias='bpt_prices',
-        post_hook='{{ expose_spells_hide_trino(\'["arbitrum"]\',
+        alias = 'bpt_prices',
+        materialized = 'incremental',
+        file_format = 'delta',
+        incremental_strategy = 'merge',
+        unique_key = ['blockchain', 'hour','contract_address'],
+        post_hook = '{{ expose_spells(\'["arbitrum"]\',
                                     "project",
                                     "balancer_v2",
-                                    \'["victorstefenon"]\') }}'
+                                    \'["victorstefenon", "thetroyharris"]\') }}'
     )
 }}
 
-WITH bpt_trades AS (
-    SELECT
-        block_time,
-        bpt_address,
-        bpt_amount_raw,
-        bpt_amount_raw / POWER(10, COALESCE(erc20a.decimals, 18)) AS bpt_amount,
-        token_amount_raw,
-        token_amount_raw / POWER(10, erc20b.decimals) AS token_amount,
-        p.price * token_amount_raw / POWER(10, erc20b.decimals) AS usd_amount
-    FROM (
-        SELECT
-            t.evt_block_time AS block_time,
-            CASE 
-                WHEN t.tokenIn = SUBSTRING(t.poolId, 0, 42) THEN t.tokenIn
-                ELSE t.tokenOut
-            END AS bpt_address,
-            CASE 
-                WHEN t.tokenIn = SUBSTRING(t.poolId, 0, 42) THEN t.amountIn
-                ELSE t.amountOut
-            END AS bpt_amount_raw,
-            CASE 
-                WHEN t.tokenIn = SUBSTRING(t.poolId, 0, 42) THEN t.tokenOut
-                ELSE t.tokenIn
-            END AS token_address,
+WITH
+    bpt_trades AS (
+        SELECT evt_tx_hash FROM {{ source('balancer_v2_arbitrum', 'Vault_evt_Swap') }} v
+        WHERE CAST(tokenIn AS VARCHAR(66)) = SUBSTRING(CAST(poolId AS VARCHAR(66)), 1, 42)
+        OR CAST(tokenOut AS VARCHAR(66)) = SUBSTRING(CAST(poolId AS VARCHAR(66)), 1, 42) 
+        {% if is_incremental() %}
+        AND v.evt_block_time >= date_trunc("day", now() - interval '1 week')
+        {% endif %} 
+    ), 
+    
+    all_trades_given_bpt_tx AS (
```

Comment:
we may look to combine `bpt_trades` and `all_trades_given_bpt_tx` CTE's into one, to be more efficient. from what i see here, it looks like the second CTE can likely be avoided all together & you could just select the fields in first CTE?

historically, joining / filtering on `tx_hash` has been inefficient, especially on it's own. this is more true for spark, but likely similar on trino/dunesql.

with that said, runtimes in gh action ci test aren't too bad still, but you could still look to optimize if you'd like

Reply:
Haha ya you are completely right. I don't remember why I wrote it that way. Fixing now.